File: Unit 07/Session 1/Problem Set 1/Problem 3: Recursive Sum.java
import java.util.List;
import java.util.Arrays;

/*
1. understand
    input: list
    output: sum of all values
    core logic: recursion

3. plan
    base case: if the list is empty, then return 0
    recursive step:
        return the value of the first element of the list + sumList(rest of the list)
*/

class RecursiveSum {

    // Time: O(n^2) because each of the n calls creates a new sublist via subList(1, size),
    //       and building that sublist takes O(k) time/space to copy for a list of length k —
    //       summing n + (n-1) + ... + 1 gives O(n^2) total work.
    // Space: O(n^2) because each call also holds a newly allocated sublist in
    //        memory until its call returns (sizes n-1, n-2, ..., 0), in addition
    //        to the O(n) call stack — the sublist copies dominate the total.
    public static int sumList(List<Integer> lst) {
        if (lst.isEmpty()) {
            return 0;
        }
        int first = lst.get(0);
        List<Integer> rest = new java.util.ArrayList<>(lst.subList(1, lst.size())); // copies remaining elements
        return first + sumList(rest);
    }

    // Public-facing overload — matches the original single-argument signature
    public static int sumList2(List<Integer> lst) {
        return sumList2(lst, 0);
    }

    // Time: O(n) because i indexes into lst and increments once per call, and
    //       lst.get(i) is an O(1) lookup with no copying — n calls, O(1) work each.
    // Space: O(n) because the call stack grows one frame per call until the
    //        base case is hit, but no extra lists are ever created — just n
    //        stacked frames, each holding a reference to the same lst and an int.
    public static int sumList2(List<Integer> lst, int i) {
        if (i == lst.size()) {
            return 0;
        }
        return lst.get(i) + sumList2(lst, i + 1);
    }

    public static void main(String[] args) {
        // Example Input: [1, 2, 3, 4, 5]
        System.out.println(sumList(Arrays.asList(1, 2, 3, 4, 5)));
        System.out.println(sumList(Arrays.asList(5)));
        System.out.println(sumList(Arrays.asList()));

        System.out.println();

        // No list copying
        System.out.println(sumList2(Arrays.asList(1, 2, 3, 4, 5)));
        System.out.println(sumList2(Arrays.asList(5)));
        System.out.println(sumList2(Arrays.asList()));
    }
}

File: Unit 07/Session 1/Problem Set 1/Problem 3: Recursive Sum.py
'''
1. understand
    input: list
    output: sum of all values
    core logic: recursion

3. plan
    base case: if the list is empty, then return 0
    recursive step:
        return the value of the first element of the list + sum_list(rest of the list)


'''

# Time: O(n^2) because because each of the n calls creates a new slice lst[1:],
#       and slicing a list of length k takes O(k) time to copy — summing
#       n + (n-1) + ... + 1 gives O(n^2) total work.
# Space: O(n^2) because each call also holds a newly allocated slice in
#        memory until its call returns (sizes n-1, n-2, ..., 0), in addition
#        to the O(n) call stack — the slice copies dominate the total.
def sum_list(lst):
    if len(lst) == 0:
        return 0
    return lst[0] + sum_list(lst[1:])

# Time: O(n) because i indexes into lst and increments once per call, and
#       lst[i] is an O(1) lookup with no copying — n calls, O(1) work each.
# Space: O(n) because the call stack grows one frame per call until the
#        base case is hit, but no extra lists are ever created — just n
#        stacked frames, each holding a reference to the same lst and an int.
def sum_list2(lst, i=0):                    # pass in a default parameter
    if len(lst) == i:
        return 0
    return lst[i] + sum_list2(lst, i + 1)   # no list iterators, slicing via indexes

# Example Input: [1, 2, 3, 4, 5]
print(sum_list([1,2,3,4,5]))
print(sum_list([5]))
print(sum_list([]))

print()

# No list iterators
print(sum_list2([1,2,3,4,5]))
print(sum_list2([5]))
print(sum_list2([]))